### PR TITLE
[Android/iOS] Swipe to close in SwipeView

### LINKED ISSUE #8767: [Bug][iOS] SwipeView SwipeBehaviorOnInvoked RemainOpen issue on iOS
### Description

Setting the `SwipeBehaviorOnInvoked` property to `RemainOpen`, after multiple `SwipeView`s have been rendered, doesn't result in the correct behavior on `SwipeView`s on iOS.

### Steps to Reproduce

1. Run the attached sample on iOS.
2. Navigate to the SwipeView behavior page.
3. Alter the SwipeBehaviorOnInvoked Picker to RemainOpen.
4. Swipe the first SwipeView from the right.
5. Swipe the second SwipeView from the right.
6. In the first SwipeView, click Delete.
7. In the second SwipeView, click Favourite.

### Expected Behavior

The first SwipeView remains open and the second SwipeView remains open.

### Actual Behavior

I get different behavior on different runs. Sometimes the first SwipeView closes. Sometimes it doesn't. Sometimes the second SwipeView closes. Sometimes it doesn't. 

The behavior I've observed the most is that the first SwipeView closes (incorrect), while the second SwipeView doesn't (correct).

### Basic Information

- Version with issue: 4.4-pre3
- IDE: VSMac 8.3.10
- Platform Target Frameworks: 
  - iOS:  13.2

### Reproduction Link

[SwipeViewDemos.zip](https://github.com/xamarin/Xamarin.Forms/files/3928299/SwipeViewDemos.zip)

### LINKED ISSUE #8778: [Bug] SwipeViewItem handler invocation intermittent issue on iOS and Android
### Description

When SwipeItems are revealed on iOS and Android (not tried UWP), clicking on them doesn't always invoke their handlers. Instead the SwipeItems close. However, this is an intermittent issue. Sometimes it works as planned. Other times it doesn't.

### Steps to Reproduce

1. Run the attached sample.
2. Browse to the SwipeView events page.
3. Swipe from the right on the Grid.
4. Click Favourite or Delete.

### Expected Behavior

An alert is displayed in response to either SwipeItem being revealed, followed by the SwipeItems closing.

### Actual Behavior

The SwipeItems close and the Grid is displayed again.

Note: this is an intermittent error. Sometimes it works as it should. Sometimes it doesn't. I can do a screen capture if required.

### Basic Information

- Version with issue: 4.4-pre3
- IDE: VSMac 8.3.10
- Platform Target Frameworks: 
  - iOS:  13.2
  - Android: 

### Reproduction Link

[SwipeViewDemos.zip](https://github.com/xamarin/Xamarin.Forms/files/3932438/SwipeViewDemos.zip)

### LINKED ISSUE #8983: [Bug] SwipeView - bad responsiveness to touch
### Description
Right now, in my opinion, SwipeView is not very useful, because it responds to touches very bad. I was testing on the emulator and on the real device, and many times the SwipeView control was not responding to touches which is very annoying. I have shown that on the GIF below.

Especially on the left side of the SwipeView content, the responsiveness is very bad.

On the real device, UX is worse than on the emulator.

Please also compare the SwipeView with the Gmail app. In the Gmail app, when I swipe the message it executes perfectly and it is much smoother than SwipeView.

I was testing only on Android.

### Steps to Reproduce

1. Open and run this project on Android on REAL DEVICE
[App3.zip](https://github.com/xamarin/Xamarin.Forms/files/3988487/App3.zip)
2. Swipe some items in the list to the left.
3. As you can the UX is not good. Sometimes the swipe occurs, sometimes not. It is a very frustrating experience for the user.

### Expected Behavior
SwipeView should work seamlessly and give the best possible UX.

### Actual Behavior
The UX is bad. The swipe is sometimes executed and sometimes not.

### Basic Information

- Version with issue: 4.4.0.991265
- IDE: Visual Studio 2019 16.4.1
- Platform Target Frameworks: 
  - Android: 9.0 (Pie)

### Screenshots

swipe_bad_responsiveness

### LINKED ISSUE #9306: [Bug] [iOS] Cannot un-reveal swipe view items on iOS / Inconsistent swipe view behaviour
### Steps to Reproduce

1. Create a swipe view using following code.
```
                <SwipeView>
                    <SwipeView.LeftItems>
                        <SwipeItems>
                            <SwipeItem Text="Favorite"
                       BackgroundColor="LightGreen" />
                            <SwipeItem Text="Delete"
                       BackgroundColor="LightPink" />
                        </SwipeItems>
                    </SwipeView.LeftItems>
                    <!-- Content -->
                    <Grid HeightRequest="60"
                        WidthRequest="300">
                        <Label Text="Swipe right"
                           HorizontalOptions="Center"
                           VerticalOptions="Center" />
                    </Grid>
                </SwipeView>
```
2.  Swipe Right on swipe view to reveal items. It should look this this.
image

3.  Swipe Left to un-reveal swipe items but it does not work on iOS. It works fine on Android.
image

### Expected Behavior
In my opinion, Android implementation is better because user could accidentally swipe on an item. I think user should be allowed to put the swipe items away without clicking on them.

### Basic Information

- Version with issue: Xamarin Forms v4.5.0 pre 2
- Platform Target Frameworks: 
  - iOS:   13.3
  - Android:  Android 9 and 10

### LINKED ISSUE #9461: [Enhancement] SwipeView Pull To Close
Is there any reason why you can't swipe to close the SwipeView similar to the context actions on a ListView?

My normal reaction to wanting to close a a context action is to swipe the other way but instead I have to tap and It took me a while to work out that I actually had to do that.

### LINKED ISSUE #9734: [Bug] SwipeView in 4.5 works only first way swipped
Using the 4.5 release .

Android

I have a swipe view inside a Collection.ItemTemplate. When you swipe left, then only left works from then on. If you go right, then only right works.

**see gif video attached**. I show that if I swipe a little you can see the left and right items appear. But if I fully swipe on way, then the other way is no longer accessable.

 
        <Grid RowSpacing="0" ColumnSpacing="0" VerticalOptions="FillAndExpand">
            <Grid.RowDefinitions>
                <RowDefinition Height="Auto" />
                <RowDefinition Height="*" />
            </Grid.RowDefinitions>

            <common:LottieLoader Grid.Row="0"  IsVisible="{Binding IsBusyAddressBookRelatedItems}" />

            <common:NoResults Grid.Row="1"
                              FontAwesomeIcon="{x:Static local:FontAwesomeFont.HandshakeAlt}"
                              Line1="There are no related items. "
                              Line2=""
                              Command="{Binding RefreshCommand}"
                              IsVisible="{Binding HasNoAddressBookRelatedItems}" />

 
            <CollectionView Grid.Row="1"
                            x:Name="MyCollectionView"
                            ItemsSource="{Binding AddressBookRelatedItems}"
                            IsVisible="{Binding HasNoAddressBookRelatedItems, Converter={ StaticResource NegateBooleanConverter }}"
                            androidSpecific:ListView.IsFastScrollEnabled="True"
                            VerticalOptions="FillAndExpand"
                            ItemSizingStrategy="MeasureAllItems">
                <CollectionView.ItemTemplate>
                    <DataTemplate x:DataType="addressBook:AddressBookRelatedItem">
                             <SwipeView > 
                                 <SwipeView.RightItems> 
                                     <SwipeItems> 
                                         <SwipeItem Command="{Binding Source={x:Reference MyPage}, Path=BindingContext.DeleteAddressBookRelatedTappedCommand}"  
                                                    CommandParameter="{Binding .}"  
                                                    Text="Delete"  
                                                    IconImageSource="{StaticResource TrashCanImage}" 
                                                    BackgroundColor="{DynamicResource DeleteButtonColor}" /> 
                                     </SwipeItems> 
                                 </SwipeView.RightItems> 
                             
                                 <SwipeView.LeftItems> 
                                     <SwipeItems> 
                                         <SwipeItem 
                                             Command="{Binding Source={x:Reference MyPage}, Path=BindingContext.EditAddressBookRelatedTappedCommand}" 
                                             CommandParameter="{Binding .}" 
                                             Text="Edit" 
                                             IconImageSource="{StaticResource EditImage}" 
                                             BackgroundColor="{DynamicResource SaveButtonColor}" /> 
                                     </SwipeItems> 
                                 </SwipeView.LeftItems> 

                            <gestures:Grid
                                TappedCommand="{Binding Source={RelativeSource AncestorType={x:Type detailModels:AddressBookDetailViewModel}},
                                Path=AddressBookRelatedTappedCommand}"
                                TappedCommandParameter="{Binding .}"
                                VerticalOptions="Start"
                                BackgroundColor="White">
                                
                                <Grid.ColumnDefinitions>
                                    <ColumnDefinition Width="Auto" />
                                    <ColumnDefinition Width="*" />
                                </Grid.ColumnDefinitions>

                                <shared:CircleImage
                                    Grid.Column="0"
                                    Source="{Binding Image}"
                                    Style="{ StaticResource AvatarXSmall }"
                                    Aspect="AspectFill"
                                    HorizontalOptions="CenterAndExpand"
                                    Margin="10,13,0,0"
                                    VerticalOptions="Start" />

                                <StackLayout
                                    Grid.Column="1"
                                    VerticalOptions="StartAndExpand"
                                    Padding="10, 10, 10, 10"
                                    Spacing="3"
                                    Orientation="Vertical"
                                    HorizontalOptions="FillAndExpand">

                                    <Label
                                        Text="{Binding Title}"
                                        HorizontalOptions="Fill"
                                        LineBreakMode="TailTruncation"
                                        MaxLines="2"
                                        VerticalOptions="EndAndExpand"
                                        Style="{DynamicResource TextSize5ListView}" />

                                    <Label
                                        Text="{Binding SubTitle}"
                                        LineBreakMode="TailTruncation"
                                        MaxLines="2"
                                        IsVisible="{ Binding IsSubTitleVisible }"
                                        VerticalOptions="FillAndExpand"
                                        Style="{DynamicResource TextSize4ListView}"
                                        FontAttributes="Bold"
                                        HorizontalOptions="Start" />

                                    <Label
                                        Text="{Binding Relation}"
                                        IsVisible="{ Binding Relation, Converter={ StaticResource IsNotStringEmptyConverter} }"
                                        LineBreakMode="WordWrap"
                                        MaxLines="2"
                                        Style="{DynamicResource LabelNoteStyle}"
                                        VerticalOptions="FillAndExpand"
                                        HorizontalOptions="Start" />

                                </StackLayout>
                            </gestures:Grid>
                           </SwipeView>  
                    </DataTemplate>
                </CollectionView.ItemTemplate>
 
            </CollectionView>

            <Grid Grid.Row="1"
                  VerticalOptions="EndAndExpand"
                  HorizontalOptions="End">
                <common:FabPageButton x:Name="fabCreateButton" Command="{Binding AddAddressBookRelatedCommand}" />
            </Grid>

        </Grid>
     
    </ContentView.Content>
   </ContentView>

animation

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

In this PR we fix several issues related to the swipe gesture in SwipeView:
- Allow swipe to close SwipeItems.
- Allow to move from one open SwipeItem to another doing swipe.
- Invalid touch in SwipeItems closing the SwipeView.

### Issues Resolved ### 

- fixes #9306
- fixes #9461
- fixes #9734
- fixes #8767
- fixes #8778
- fixes #8983 

### API Changes ###
 
 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

None

#### Before
wrong-behavior

Realize that to close the SwipeView the tap works but not the swipe. In a similar way, trying to move from the left open SwipeItem to the right one was not possible without previously closing the SwipeView.

#### After

Android
swipe-close-droid
swipe-items-droid

iOS
swipe-close-ios
swipe-items-ios

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs [hunk 1]
```diff
@@ -25,17 +26,28 @@ public Issue8781()
...
-		{
-			await DisplayAlert("Incorrect!", "Try again.", "OK");
-		}
+		public ICommand CheckAnswerCommand => new Command<string>(OnCheckAnswer);
 
-		async void OnCorrectAnswerInvoked(object sender, EventArgs e)
+		async void OnCheckAnswer(string parameter)
 		{
-			await DisplayAlert("Correct!", "The answer is 2.", "OK");
+			if (string.IsNullOrEmpty(parameter))
```
Comment (jfversluis) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs]:
```suggestion
			#if APP
			if (string.IsNullOrEmpty(parameter))
```

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs [hunk 2]
```diff
@@ -25,17 +26,28 @@ public Issue8781()
...
+				resultEntry.Text = string.Empty;
+				swipeView.Close();
+				await DisplayAlert("Correct!", "The answer is 2.", "OK");
+			}
+			else
+			{
+				resultEntry.Text = string.Empty;
+				await DisplayAlert("Incorrect!", "Try again.", "OK");
+			}
 		}
```
Comment (jfversluis) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs]:
```suggestion
				#endif
		}
```

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs (hunk 1 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs]: Right. Thanks for the feedback. Updated this repro sample to fix the build.

**Comment by jfversluis** [comment]
> ##[error]Xamarin.Forms.Controls.Issues\Xamarin.Forms.Controls.Issues.Shared\Issue8781.xaml.cs(37,14): Error CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.

**Comment by vniehues** [comment]
In Previous versions the SwipeView was preventing underlying TapGestureRecognizers from working; 
e.g. You have a list with swipeViews but you also want to open a detail page using a simple tap. that tap didn't come through.

Is that fixed as well? Or should I open a new issue for that one?

**Comment by jsuarezruiz** [comment]
@vniehues Could you take a look to this PR https://github.com/xamarin/Xamarin.Forms/pull/9638?, do you mean the selection events?

**Comment by vniehues** [comment]
@jsuarezruiz In the PR and the gif you showed it looks like my issue will be resolved.

~~Just to clarify, I don't mean the built-in selection event. In my case I added a TapGestureRecognizer to the ItemTemplates first child (StackLayout in this case) but the TapGestureRecognizer did not get the Input. At least it didn't call the specified command (which it did before adding the SwipeView).~~

Do I have to add the TapGestureRecognizer to the SwipeView or its Content (e.g. StackLayout)?

**Comment by jsuarezruiz** [comment]
@vniehues Using the Content will work.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs [hunk 3]
```diff
@@ -24,18 +26,31 @@ public partial class Issue8781 : ContentPage
...
+				return;
 
-		async void OnCorrectAnswerInvoked(object sender, EventArgs e)
-		{
-			await DisplayAlert("Correct!", "The answer is 2.", "OK");
+			if (parameter.Equals("4", StringComparison.InvariantCultureIgnoreCase))
+			{
+				resultEntry.Text = string.Empty;
+				swipeView.Close();
+				await DisplayAlert("Correct!", "The answer is 2.", "OK");
```
Comment (PureWeen) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs]: Is it though :-p

Comment (jsuarezruiz) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue8781.xaml.cs]: Haha, fixed!

**Comment by jsuarezruiz** [comment]
Thanks for the feedback! I have added more samples and UITests to verify the SwipView opening and closing by doing swipe.

**Pull request merged**